File: sample/client/py/engine/engine/conn_msg_handle.py
# -*- coding: UTF-8 -*-
from collections.abc import Callable
from .msgpack import *

class conn_msg_handle(object):
    def on_create_remote_entity(self, entity_type:str, entity_id:str, argvs:bytes):
        from app import app
        app().create_entity(entity_type, entity_id, loads(argvs))

    def on_delete_remote_entity(self, entity_id:str):
        from app import app
        app().delete_entity(entity_id)
        
    def on_refresh_entity(self, entity_type:str, entity_id:str, argvs:bytes):
        from app import app
        app().update_entity(entity_type, entity_id, loads(argvs))
        
    def on_kick_off(self, prompt_info:str):
        from app import app
        app().close()
        app().on_kick_off(prompt_info)

    def on_transfer_complete(self):
        from app import app
        app().on_transfer_complete()
        
    def on_call_rpc(self, hub_name:str, entity_id:str, msg_cb_id:int, method:str, argvs:bytes):
        from app import app
        _player = app().player_mgr.get_player(entity_id)
        if _player != None:
            _player.handle_hub_request(method, hub_name, msg_cb_id, argvs)
            return
        print("unhandle hub request method:{} entity:{}".format(method, entity_id))

    def on_call_rsp(self, entity_id:str, msg_cb_id:int, argvs:bytes):
        from app import app
        _player = app().player_mgr.get_player(entity_id)
        if _player != None:
            _player.handle_hub_response(msg_cb_id, argvs)
            return
        _subentity = app().subentity_mgr.get_subentity(entity_id)
        if _subentity != None:
            _subentity.handle_hub_response(msg_cb_id, argvs)
            return
        print("unhandle hub response msg_cb_id:{} entity:{}".format(msg_cb_id, entity_id))

    def on_call_err(self, entity_id:str, msg_cb_id:int, argvs:bytes):
        from app import app
        _player = app().player_mgr.get_player(entity_id)
        if _player != None:
            _player.handle_hub_response_error(msg_cb_id, argvs)
            return
        _subentity = app().subentity_mgr.get_subentity(entity_id)
        if _subentity != None:
            _subentity.handle_hub_response_error(msg_cb_id, argvs)
            return
        print("unhandle hub response error msg_cb_id:{} entity:{}".format(msg_cb_id, entity_id))

    def on_call_ntf(self, hub_name:str, entity_id:str, method:str, argvs:bytes):
        from app import app
        _player = app().player_mgr.get_player(entity_id)
        if _player != None:
            _player.handle_hub_notify(method, hub_name, argvs)
            return
        _subentity = app().subentity_mgr.get_subentity(entity_id)
        if _subentity != None:
            _subentity.handle_hub_notify(method, hub_name, argvs)
            return
        _receiver = app().receiver_mgr.get_receiver(entity_id)
        if _receiver != None:
            _receiver.handle_hub_notify(method, hub_name, argvs)
            return
        print("unhandle hub response notify msg_cb_id:{} entity:{}".format(method, entity_id))
        
    def on_call_global(self, method:str, argvs:bytes):
        from app import app
        app().on_call_global(method, argvs)

File: sample/client/ts/engine/engine/conn_msg_handle.ts
/*
 * conn_msg_handle.ts
 * qianqians
 * 2023/10/5
 */
import { decode } from '../@msgpack/msgpack'
import * as app2 from './app'

export class conn_msg_handle {
    public on_create_remote_entity(entity_type:string, entity_id:string, argvs:Uint8Array) {
        let info = decode<object>(argvs);
        if (typeof info === "object" && info) {
            app2.app.instance.create_entity(entity_type, entity_id, info);
        }
    }
    
    public on_delete_remote_entity(entity_id:string) {
        app2.app.instance.delete_entity(entity_id);
    }
    
    public on_refresh_entity(entity_type:string, entity_id:string, argvs:Uint8Array) {
        let info = decode<object>(argvs);
        if (typeof info === "object" && info) {
            app2.app.instance.update_entity(entity_type, entity_id, info);
        }
    }

    public on_kick_off(prompt_info:string) {
        app2.app.instance.on_kick_off(prompt_info);
    }
    
    public on_transfer_complete() {
        app2.app.instance.on_transfer_complete();
    }
    
    public on_call_rpc(hub_name:string, entity_id:string, msg_cb_id:number, method:string, argvs:Uint8Array) {
        let _player = app2.app.instance.player_mgr.get_player(entity_id);
        if (_player) {
            _player.handle_hub_request(method, hub_name, msg_cb_id, argvs);
        }
    }
    
    public on_call_rsp(entity_id:string, msg_cb_id:number, argvs:Uint8Array) {
        let _player = app2.app.instance.player_mgr.get_player(entity_id);
        if (_player) {
            _player.handle_hub_response(msg_cb_id, argvs);
            return;
        }

        let _subentity = app2.app.instance.subentity_mgr.get_subentity(entity_id);
        if (_subentity) {
            _subentity.handle_hub_response(msg_cb_id, argvs);
            return;
        }
    }
    
    public on_call_err(entity_id:string, msg_cb_id:number, argvs:Uint8Array) {
        let _player = app2.app.instance.player_mgr.get_player(entity_id);
        if (_player) {
            _player.handle_hub_response_error(msg_cb_id, argvs);
            return;
        }

        let _subentity = app2.app.instance.subentity_mgr.get_subentity(entity_id);
        if (_subentity) {
            _subentity.handle_hub_response_error(msg_cb_id, argvs);
            return;
        }
    }
    
    public on_call_ntf(hub_name:string, entity_id:string, method:string, argvs:Uint8Array) {
        let _player = app2.app.instance.player_mgr.get_player(entity_id);
        if (_player) {
            _player.handle_hub_notify(method, hub_name, argvs);
            return;
        }

        let _subentity = app2.app.instance.subentity_mgr.get_subentity(entity_id);
        if (_subentity) {
            _subentity.handle_hub_notify(method, hub_name, argvs);
            return;
        }

        let _receiver = app2.app.instance.receiver_mgr.get_receiver(entity_id)
        if (_receiver) {
            _receiver.handle_hub_notify(method, hub_name, argvs);
            return
        }
    }
    
    public on_call_global(method:string, hub_name:string, argvs:Uint8Array) {
        app2.app.instance.on_call_global(method, hub_name, argvs);
    }
}